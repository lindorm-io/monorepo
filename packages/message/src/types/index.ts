export * from "./decorators";
export * from "./global";
export * from "./message-kit";
export * from "./metadata";
export * from "./scanner";
export * from "./subscription";
