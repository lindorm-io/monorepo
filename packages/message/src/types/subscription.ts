import { IMessage } from "../interfaces";

export type SubscribeOptions<M extends IMessage = IMessage> = {
  callback(message: M): Promise<void>;
  consumerTag?: string;
  topic: string;
  queue?: string;
};

export type FindSubscriptionFilter = {
  queue: string;
  topic: string;
};

export type RemoveSubscriptionFilter = {
  consumerTag: string;
};

export type UnsubscribeOptions = {
  topic: string;
  queue: string;
};
