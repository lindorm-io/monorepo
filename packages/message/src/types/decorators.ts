import { ZodType } from "zod";
import { MetaFieldFallback, MetaFieldType, MetaGeneratedStrategy } from "./metadata";

export type FieldDecoratorOptions = {
  enum?: any;
  fallback?: MetaFieldFallback;
  max?: number;
  min?: number;
  nullable?: boolean;
  optional?: boolean;
  schema?: ZodType;
  type?: MetaFieldType;
};

export type MessageDecoratorOptions = {
  name?: string;
  namespace?: string;
  topic?: string;
};

export type GeneratedDecoratorOptions = {
  strategy?: MetaGeneratedStrategy;
  length?: number;
  max?: number;
  min?: number;
};

export type HookDecoratorCallback = (message: any) => void;
