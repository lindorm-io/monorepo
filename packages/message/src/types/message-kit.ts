import { ILogger } from "@lindorm/logger";
import { Constructor } from "@lindorm/types";
import { IMessage } from "../interfaces";

export type MessageKitOptions<M extends IMessage> = {
  Message: Constructor<M>;
  logger: ILogger;
};

export type TopicNameOptions = {
  namespace?: string | null;
  topic?: string | null;
};
