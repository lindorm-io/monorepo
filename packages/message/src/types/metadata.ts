/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { Dict } from "@lindorm/types";
import { ZodObject, ZodType } from "zod";
import { IMessage } from "../interfaces";

export type MetaFieldType =
  | "array"
  | "bigint"
  | "boolean"
  | "date"
  | "email"
  | "enum"
  | "float"
  | "integer"
  | "object"
  | "string"
  | "url"
  | "uuid";

export type MetaFieldDecorator =
  // default
  | "Field"
  // special (unique)
  | "CorrelationField"
  | "DelayField"
  | "IdentifierField"
  | "MandatoryField"
  | "PersistentField"
  | "PriorityField"
  // date
  | "TimestampField";

export type FieldFallbackPrimitive = bigint | boolean | Date | number | string | null;
export type FieldFallbackComplex =
  | Array<FieldFallbackPrimitive>
  | Dict<FieldFallbackPrimitive>
  | FieldFallbackPrimitive;
export type MetaFieldFallback = (() => FieldFallbackComplex) | FieldFallbackComplex;

export type MetaField<T extends MetaFieldDecorator = MetaFieldDecorator> = {
  target: Function;
  key: string;
  decorator: T;
  enum: any | null;
  fallback: MetaFieldFallback | null;
  max: number | null;
  min: number | null;
  nullable: boolean;
  optional: boolean;
  schema: ZodType | null;
  type: MetaFieldType | null;
};

export type MetaMessage = {
  target: Function;
  decorator: string;
  name: string | null;
  namespace: string | null;
  topic: string | null;
};

export type MetaGeneratedStrategy = "date" | "float" | "integer" | "string" | "uuid";

export type MetaGenerated = {
  target: Function;
  key: string;
  length: number | null;
  max: number | null;
  min: number | null;
  strategy: MetaGeneratedStrategy;
};

export type MetaHookDecorator =
  // before
  | "OnCreate"
  | "OnValidate"
  // after
  | "OnPublish"
  | "OnConsume";

export type MetaHook = {
  target: Function;
  decorator: MetaHookDecorator;
  callback: (message: any) => void;
};

export type MetaSchema = {
  target: Function;
  schema: ZodObject<IMessage>;
};

export type MessageMetadata<T extends MetaFieldDecorator = MetaFieldDecorator> = {
  fields: Array<Omit<MetaField<T>, "target">>;
  generated: Array<Omit<MetaGenerated, "target">>;
  hooks: Array<Omit<MetaHook, "target">>;
  message: Omit<MetaMessage, "target">;
  schemas: Array<ZodObject<IMessage>>;
};
