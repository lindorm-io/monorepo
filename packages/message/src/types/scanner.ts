import { Constructor, Dict } from "@lindorm/types";

export type MessageScannerInput<T extends Dict = Dict> = Array<Constructor<T> | string>;
