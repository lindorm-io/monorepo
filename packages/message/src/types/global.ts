/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { GlobalMessageMetadata as GlobalMessageMetadataImpl } from "../classes/private";
import { MetaField, MetaGenerated, MetaHook, MetaMessage, MetaSchema } from "./metadata";

export type GlobalMessageMetadata = {
  fields: Array<MetaField>;
  messages: Array<MetaMessage>;
  generated: Array<MetaGenerated>;
  hooks: Array<MetaHook>;
  schemas: Array<MetaSchema>;
};

export type GlobalThisMessage = typeof globalThis & {
  __lindorm_message__: GlobalMessageMetadataImpl;
};
