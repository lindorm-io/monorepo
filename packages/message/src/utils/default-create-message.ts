import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "./global";
import { parseField } from "./private";

export const defaultCreateMessage = <
  E extends IMessage,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Message: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalMessageMetadata.get(Message);
  const message = new Message();

  for (const field of metadata.fields) {
    (message as any)[field.key] = parseField(field, message, options);

    if ((message as any)[field.key]) continue;

    (message as any)[field.key] = isFunction(field.fallback)
      ? field.fallback()
      : field.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(message);
  }

  return message;
};
