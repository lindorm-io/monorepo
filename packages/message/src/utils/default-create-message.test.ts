import { Dict } from "@lindorm/types";
import MockDate from "mockdate";
import z from "zod";
import {
  CorrelationField,
  DelayField,
  Field,
  IdentifierField,
  MandatoryField,
  Message,
  OnConsume,
  OnCreate,
  OnPublish,
  OnValidate,
  PersistentField,
  PriorityField,
  Schema,
  TimestampField,
} from "../decorators";
import { defaultCreateMessage } from "./default-create-message";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomUUID: jest
    .fn()
    .mockImplementation(() => "c8ff3952-8ba4-51a3-a4d5-2f0726c49524")
    .mockImplementationOnce(() => "e8edbe03-f52f-56c6-8d98-82e6961329fe")
    .mockImplementationOnce(() => "0c8c31df-f1cc-5381-97fc-31b3714327de")
    .mockImplementationOnce(() => "2dc92d72-9ce3-50c7-b982-59022ec393f1")
    .mockImplementationOnce(() => "88943033-fef1-5ad8-8cf8-4470be80d197")
    .mockImplementationOnce(() => "1d37a07d-4fd6-5913-85eb-2c7d721cf2a9")
    .mockImplementationOnce(() => "5f2976d2-afda-5e28-9c06-3c04865ea8df"),
}));

describe("defaultCreateMessage", () => {
  enum TestEnum {
    One = "1",
    Two = "2",
  }

  test("should create with empty values", () => {
    @Message()
    @Schema(z.object({}).passthrough())
    @OnCreate((message) => {
      message.OnCreate = "OnCreate";
    })
    @OnConsume((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnPublish((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnValidate((message) => {
      message.OnValidate = "OnValidate";
    })
    class TestCreateMessageEmpty {
      @CorrelationField()
      correlation!: string;

      @DelayField()
      delay!: number;

      @IdentifierField()
      identifier!: string;

      @MandatoryField()
      mandatory!: boolean;

      @PersistentField()
      persistent!: boolean;

      @PriorityField()
      priority!: number;

      @TimestampField()
      timestamp!: Date;

      @Field("array")
      array!: any[];

      @Field("bigint")
      bigint!: bigint;

      @Field("boolean")
      boolean!: boolean;

      @Field("date")
      date!: Date;

      @Field("email")
      email!: string;

      @Field("enum", { enum: TestEnum })
      enum!: TestEnum;

      @Field("float")
      float!: number;

      @Field("integer")
      integer!: number;

      @Field("object")
      object!: Dict;

      @Field("string")
      string!: string;

      @Field("url")
      url!: string;

      @Field("uuid")
      uuid!: string;
    }

    expect(defaultCreateMessage(TestCreateMessageEmpty)).toMatchSnapshot();
  });

  test("should create with fallback values", () => {
    @Message()
    @Schema(z.object({}).passthrough())
    @OnCreate((message) => {
      message.OnCreate = "OnCreate";
    })
    @OnConsume((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnPublish((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnValidate((message) => {
      message.OnValidate = "OnValidate";
    })
    class TestCreateMessageFallback {
      @CorrelationField()
      correlation!: string;

      @DelayField()
      delay!: number;

      @IdentifierField()
      identifier!: string;

      @MandatoryField()
      mandatory!: boolean;

      @PersistentField()
      persistent!: boolean;

      @PriorityField()
      priority!: number;

      @TimestampField()
      timestamp!: Date;

      @Field("array", { fallback: () => [] })
      array!: any[];

      @Field("bigint", { fallback: () => BigInt(0) })
      bigint!: bigint;

      @Field("boolean", { fallback: () => false })
      boolean!: boolean;

      @Field("date", { fallback: () => new Date("2024-01-01T08:00:00.000Z") })
      date!: Date;

      @Field("email", { fallback: () => "ba@fer.af" })
      email!: string;

      @Field("enum", { enum: TestEnum, fallback: () => TestEnum.One })
      enum!: TestEnum;

      @Field("float", { fallback: () => 0.0 })
      float!: number;

      @Field("integer", { fallback: () => 0 })
      integer!: number;

      @Field("object", { fallback: () => ({ object: true }) })
      object!: Dict;

      @Field("string", { fallback: () => "string" })
      string!: string;

      @Field("url", { fallback: () => "https://example.com" })
      url!: string;

      @Field("uuid", { fallback: () => "f264a693-99a3-5cb0-955c-f70fd90e59d4" })
      uuid!: string;
    }

    expect(defaultCreateMessage(TestCreateMessageFallback)).toMatchSnapshot();
  });

  test("should create with options values", () => {
    @Message()
    @Schema(z.object({}).passthrough())
    @OnCreate((message) => {
      message.OnCreate = "OnCreate";
    })
    @OnConsume((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnPublish((message) => {
      message.OnValidate = "OnValidate";
    })
    @OnValidate((message) => {
      message.OnValidate = "OnValidate";
    })
    class TestCreateMessageOptions {
      @CorrelationField()
      correlation!: string;

      @DelayField()
      delay!: number;

      @IdentifierField()
      identifier!: string;

      @MandatoryField()
      mandatory!: boolean;

      @PersistentField()
      persistent!: boolean;

      @PriorityField()
      priority!: number;

      @TimestampField()
      timestamp!: Date;

      @Field("array")
      array!: any[];

      @Field("bigint")
      bigint!: bigint;

      @Field("boolean")
      boolean!: boolean;

      @Field("date")
      date!: Date;

      @Field("email")
      email!: string;

      @Field("enum", { enum: TestEnum })
      enum!: TestEnum;

      @Field("float")
      float!: number;

      @Field("integer")
      integer!: number;

      @Field("object")
      object!: Dict;

      @Field("string")
      string!: string;

      @Field("url")
      url!: string;

      @Field("uuid")
      uuid!: string;
    }

    expect(
      defaultCreateMessage(TestCreateMessageOptions, {
        correlation: "b3f411b3-aafc-5445-b2f4-b761c8dce43e",
        delay: 1234,
        identifier: "9350aeb8-8bb2-57f0-9268-64f955683261",
        mandatory: true,
        persistent: true,
        priority: 999,
        timestamp: new Date("2025-01-01T08:00:00.000Z"),
        array: ["array", "of", "values"],
        bigint: BigInt(1234567890123456789),
        boolean: true,
        date: new Date("2026-01-01T08:00:00.000Z"),
        email: "sus@uco.as",
        enum: TestEnum.Two,
        float: 3.14,
        integer: 42,
        object: { key: "value" },
        string: "x7cyJh9R3w1",
        url: "http://comomzi.kh/iwibavif",
        uuid: "c6263d16-9031-571c-92c1-f62f5dec8286",
      }),
    ).toMatchSnapshot();
  });
});
