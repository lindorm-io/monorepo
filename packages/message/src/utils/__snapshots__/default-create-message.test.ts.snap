// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultCreateMessage should create with empty values 1`] = `
TestCreateMessageEmpty {
  "OnCreate": "OnCreate",
  "array": null,
  "bigint": null,
  "boolean": null,
  "correlation": "e8edbe03-f52f-56c6-8d98-82e6961329fe",
  "date": null,
  "delay": 0,
  "email": null,
  "enum": null,
  "float": null,
  "identifier": "0c8c31df-f1cc-5381-97fc-31b3714327de",
  "integer": null,
  "mandatory": false,
  "object": null,
  "persistent": false,
  "priority": 1,
  "string": null,
  "timestamp": 2024-01-01T08:00:00.000Z,
  "url": null,
  "uuid": null,
}
`;

exports[`defaultCreateMessage should create with fallback values 1`] = `
TestCreateMessageFallback {
  "OnCreate": "OnCreate",
  "array": [],
  "bigint": 0n,
  "boolean": false,
  "correlation": "2dc92d72-9ce3-50c7-b982-59022ec393f1",
  "date": 2024-01-01T08:00:00.000Z,
  "delay": 0,
  "email": "ba@fer.af",
  "enum": "1",
  "float": 0,
  "identifier": "88943033-fef1-5ad8-8cf8-4470be80d197",
  "integer": 0,
  "mandatory": false,
  "object": {
    "object": true,
  },
  "persistent": false,
  "priority": 1,
  "string": "string",
  "timestamp": 2024-01-01T08:00:00.000Z,
  "url": "https://example.com",
  "uuid": "f264a693-99a3-5cb0-955c-f70fd90e59d4",
}
`;

exports[`defaultCreateMessage should create with options values 1`] = `
TestCreateMessageOptions {
  "OnCreate": "OnCreate",
  "array": [
    "array",
    "of",
    "values",
  ],
  "bigint": 1234567890123456768n,
  "boolean": true,
  "correlation": "b3f411b3-aafc-5445-b2f4-b761c8dce43e",
  "date": 2026-01-01T08:00:00.000Z,
  "delay": 1234,
  "email": "sus@uco.as",
  "enum": "2",
  "float": 3.14,
  "identifier": "9350aeb8-8bb2-57f0-9268-64f955683261",
  "integer": 42,
  "mandatory": true,
  "object": {
    "key": "value",
  },
  "persistent": true,
  "priority": 999,
  "string": "x7cyJh9R3w1",
  "timestamp": 2025-01-01T08:00:00.000Z,
  "url": "http://comomzi.kh/iwibavif",
  "uuid": "c6263d16-9031-571c-92c1-f62f5dec8286",
}
`;
